docs/puzzles/day16.md

```diff
@@ -1,11 +1,572 @@
+    (toLong(bits), rest)
+end readLiteralBody
+```
+
+In each step we read 5 bits from the input and check if we should finish. If the
+first bit is `0` then we know that we can append the last 5 bits and return the
+current result. In case the first bit is `1`, we need to repeat the step once
+more on the remaining bits.
+
+The harder part will be defining `readOperatorBody` since we know that operator
+packets can contain other packets and those packets can also be operators! This
+means we will need to apply a recursive approach:
+
+```scala
+def readOperatorBody(current: BinaryData): (List[Packet], BinaryData) =
+  val (lenId, rest) = current.splitAt(1)
+
+  @tailrec
+  def readMaxBits(
+      current: BinaryData,
+      remaining: Int,
+      acc: List[Packet]
+  ): (List[Packet], BinaryData) =
+    if remaining == 0 then (acc, current)
+    else
+      val (newExpr, rest) = decodePacket(current)
+      readMaxBits(rest, remaining - (current.size - rest.size), acc :+ newExpr)
+
+  @tailrec
+  def readMaxPackets(
+      current: BinaryData,
+      remaining: Int,
+      acc: List[Packet]
+  ): (List[Packet], BinaryData) =
+    if remaining == 0 then (acc, current)
+    else
+      val (newExpr, rest) = decodePacket(current)
+      readMaxPackets(rest, remaining - 1, acc :+ newExpr)
+
+  // read based on length
+  if lenId(0) == '0' then
+    val (size, packets) = rest.splitAt(15)
+    readMaxBits(packets, toInt(size), Nil)
+  // read based on number of packages
+  else
+    val (size, packets) = rest.splitAt(11)
+    readMaxPackets(packets, toInt(size), Nil)
+  end match
+end readOperatorBody
+```
+
+In the above function we first check the first bit of the operator body, which
+tell us how we should check the rest of the body.
+
+- if the bit is `0` it means that the next 15 bits can be turned into a number,
+  that will define how many of the further bits are the subpackets of the
+  operator.
+
+- if the bit is `1` it means that the next 11 bits can be turned into a number,
+  that will defining how many subpackets should belong to the operator.
+
+We defined two helper recursive functions `readMaxBits` and `readMaxPackets`
+which will check if the stopping condition (either max bits read or max
+packets read) is achieved or read a new packet using recursively the
+`decodePacket` function otherwise. At the end they will both return a list of
+packets, that we can later use to put into the operator packet, and the
+remaining input that we might need to check for more packets.
+
+This should already allow us to create a full structure and what remains is
+adding a function that can sum all the versions up. We can just add that
+function to the `Packet` enum and sum it all recursively.
+
+```scala
+  def versionSum: Int =
+    this match
+      case Literal(version, _)     => version
+      case Operator(version, exprs, _) => version + exprs.map(_.versionSum).sum
+```
+
+That's it! We should be able to solve the part 1.
+
+### Full solution
```

Comment:
Should I maybe just add the full solution at the end?

Reply:
I personally like seeing first the solution of part1

Reply:
This seems fine, since it is the `Full solution` of part 1.

Reply:
ok cool!